File: lib/features/subscription/domain/entitlement.dart
/// Domain model for the subscription entitlement.
///
/// Mirrors the RevenueCat entitlement identifier (`pro`). RevenueCat is the
/// only gate (mp-279): the client reads the SDK's cached entitlement and
/// nothing else, so there is one source here, not three. Pure Dart, no SDK
/// or Supabase types.
library;

import 'grant.dart';

/// Entitlements the app knows how to gate on. The [key] is the identifier in
/// RevenueCat; the server's own two-field cache (mp-285) keys on the user.
enum Entitlement {
  pro('pro');

  const Entitlement(this.key);

  /// RevenueCat entitlement identifier.
  final String key;
}

/// How long past its own [SubscriptionStatus.expiresAt] an active answer
/// still counts (mp-679): a saved copy on the phone keeps the Gate open this
/// long while the fetch for the renewal runs, then counts as closed until a
/// fresh answer arrives.
///
/// KEEP IN STEP with `RENEWAL_GRACE_MS` in
/// supabase/functions/_shared/vana/entitlement.ts, the server's grace for a
/// renewing account (15 minutes): the two must move together, or the phone
/// opens onto a server that refuses every AI call.
const Duration kRenewalGrace = Duration(minutes: 15);

/// Which source vouched for the active entitlement. [none] when inactive.
enum SubscriptionSource { none, revenuecat }

/// The gate's answer (mp-457): open or closed, nothing in between.
enum AppAccess {
  /// The entitlement is active, or the account is a team admin (mp-416).
  open,

  /// No live `pro`: never held, held once and expired, or an unknown answer
  /// (mp-284). The full-screen paywall and nothing else (mp-280, mp-611).
  closed;

  /// Whether AI actions may run: only an open gate.
  bool get allowsAi => this == open;
}

/// The resolved subscription status for the current user.
class SubscriptionStatus {
  const SubscriptionStatus({
    required this.active,
    this.expiresAt,
    this.source = SubscriptionSource.none,
    this.isTrial = false,
    this.productId,
    this.willRenew = true,
    this.hadPro = false,
    this.grant,
  });

  /// Nobody is subscribed (also the safe fallback whenever a lookup fails
  /// or times out: an unknown entitlement is locked, mp-284).
  static const none = SubscriptionStatus(active: false);

  /// Whether the app is unlocked for this user right now.
  final bool active;

  /// When the current period ends (UTC). Null for open-ended grants and for
  /// [none]. A past [expiresAt] with [active] true (RevenueCat's saved copy
  /// judges itself against the time it was fetched) counts only for
  /// [kRenewalGrace]: see [countedAt].
  final DateTime? expiresAt;

  /// Who said so. [SubscriptionSource.none] when [active] is false.
  final SubscriptionSource source;

  /// True while the entitlement is in a trial (or intro) period.
  final bool isTrial;

  /// The store SKU that granted the entitlement, when known.
  final String? productId;

  /// Whether the store will renew (or, in a trial, start charging) at
  /// [expiresAt]. False once the athlete has cancelled: the entitlement
  /// stays active to the end of the period, then lapses. The day-five
  /// reminder is cancelled on an active trial that will not renew (mp-456).
  final bool willRenew;

  /// Whether RevenueCat has ever recorded `pro` for this customer: true
  /// while it is active and after it has expired, false for a customer who
  /// never held it. The gate does not read it: lapsed and never are both
  /// closed (mp-457, mp-611). An unknown answer ([none]) is false.
  final bool hadPro;

  /// The Grant behind an active `pro`, when RevenueCat granted it rather
  /// than a store selling it (mp-558); null otherwise.
  final Grant? grant;

  /// Whether [expiresAt] has passed as of [now]. False when there is no expiry.
  bool isExpiredAt(DateTime now) {
    final e = expiresAt;
    return e != null && !e.isAfter(now);
  }

  /// The moment this answer stops counting (mp-679): its own [expiresAt],
  /// plus [kRenewalGrace] when it will renew. Like the server's isEntitled,
  /// only a renewing subscription gets the grace; a cancelled one ends at
  /// its expiry. Null for an inactive answer and for an open-ended one.
  DateTime? get stopsCountingAt {
    final e = expiresAt;
    if (!active || e == null) return null;
    return willRenew ? e.add(kRenewalGrace) : e;
  }

  /// This answer as it counts at [now] (mp-679). An active answer whose own
  /// [expiresAt] passed more than [kRenewalGrace] ago (at once, when it will
  /// not renew) is closed: held once,
  /// same expiry and product, nothing vouching for it. Every other answer
  /// counts as it is, so a fresh answer, which carries the new period's
  /// expiry, always wins.
  SubscriptionStatus countedAt(DateTime now) {
    final e = expiresAt;
    final end = stopsCountingAt;
    if (e == null || end == null || now.isBefore(end)) return this;
    return SubscriptionStatus(
      active: false,
      expiresAt: e,
      productId: productId,
      willRenew: false,
      hadPro: true,
    );
  }

  SubscriptionStatus copyWith({
    bool? active,
    DateTime? expiresAt,
    SubscriptionSource? source,
    bool? isTrial,
    String? productId,
    bool? willRenew,
    bool? hadPro,
    Grant? grant,
  }) {
    return SubscriptionStatus(
      active: active ?? this.active,
      expiresAt: expiresAt ?? this.expiresAt,
      source: source ?? this.source,
      isTrial: isTrial ?? this.isTrial,
      productId: productId ?? this.productId,
      willRenew: willRenew ?? this.willRenew,
      hadPro: hadPro ?? this.hadPro,
      grant: grant ?? this.grant,
    );
  }

  @override
  bool operator ==(Object other) =>
      other is SubscriptionStatus &&
      other.active == active &&
      other.expiresAt == expiresAt &&
      other.source == source &&
      other.isTrial == isTrial &&
      other.productId == productId &&
      other.willRenew == willRenew &&
      other.hadPro == hadPro &&
      other.grant == grant;

  @override
  int get hashCode => Object.hash(
    active,
    expiresAt,
    source,
    isTrial,
    productId,
    willRenew,
    hadPro,
    grant,
  );

  @override
  String toString() =>
      'SubscriptionStatus(active: $active, source: ${source.name}, '
      'expiresAt: $expiresAt, isTrial: $isTrial, productId: $productId, '
      'willRenew: $willRenew, hadPro: $hadPro, grant: $grant)';
}

/// A free introductory period the store attaches to a subscription product
/// (mp-279: seven days free on the monthly and annual plans). Only a free
/// offer is modelled — a discounted intro price is not one this app sells.
class IntroOffer {
  const IntroOffer({required this.freeDays});

  /// Length of the free period in days.
  final int freeDays;

  /// Days for a store period of [count] × [unit], where [unit] is the
  /// store's DAY / WEEK / MONTH / YEAR word (case-insensitive). Apple reports
  /// a seven-day trial as `WEEK × 1`; the copy says "7 days", so weeks are
  /// unrolled. Months and years use the 30 / 365 convention the stores use
  /// for their own copy. Unknown units yield null.
  static int? daysFor({required String unit, required int count}) {
    if (count <= 0) return null;
    return switch (unit.toUpperCase()) {
      'DAY' => count,
      'WEEK' => count * 7,
      'MONTH' => count * 30,
      'YEAR' => count * 365,
      _ => null,
    };
  }

  @override
  bool operator ==(Object other) =>
      other is IntroOffer && other.freeDays == freeDays;

  @override
  int get hashCode => freeDays.hashCode;

  @override
  String toString() => 'IntroOffer(freeDays: $freeDays)';
}

File: supabase/functions/_shared/vana/entitlement.ts
/**
 * The one server gate (mp-266, mp-279, mp-285): a caller passes when their
 * `public.user_entitlements` row — the small cache of RevenueCat that only
 * the revenuecat-webhook writes — has `active_until` in the future, or, while
 * the row says the subscription renews, less than [RENEWAL_GRACE_MS] in the
 * past (mp-457: a paying account keeps Pro between a period end and a late
 * RENEWAL webhook; testing-wave 06-002).
 *
 * There is no flag, no trial clock and no tester bypass on this path: nothing
 * app-side can grant an entitlement. Testers get a RevenueCat promotional
 * grant, which arrives through the webhook like any other event.
 *
 * One exception, ruled 2026-09-26 (testing-wave 122-004, superseding
 * mp-416's server check for admins): a team admin, `public.users.is_admin`
 * set by hand in the database (mp-144 clause 3), gets every feature. The flag
 * is read only when the cache says no, and only a literal `true` counts.
 */
import type { Db } from './env.ts';
import { jsonResponse } from '../responses.ts';

export type ProCheck = { ok: true } | { ok: false; reason: 'pro_required' };

/**
 * How long past `active_until` a renewing subscription keeps Pro on the server
 * while its RENEWAL webhook is on the way. Test Store renewals landed 2.5 and
 * 4 minutes after the period end (06-002); RevenueCat's first two webhook
 * retries come 5 and 10 minutes after a failed delivery. A cancelled
 * subscription, a grant or a closed row gets no grace.
 *
 * KEEP IN STEP with `public.entitlement_renewal_grace()`
 * (supabase/migrations/20260925110000_allowance_survives_renewal_grace.sql),
 * which keeps the monthly allowance alive for the same grace: the two must
 * move together, or the gate opens onto an empty wallet.
 */
export const RENEWAL_GRACE_MS = 15 * 60_000;

/** The row as the webhook writes it (period_type is carried for callers, not read here). */
export interface EntitlementCacheRow {
  active_until: string | null;
  period_type?: string | null;
  /** RevenueCat's last word that the store subscription renews at `active_until`. */
  will_renew?: boolean | null;
}

/**
 * Pure rule: access runs while `active_until` is later than now, and for
 * [RENEWAL_GRACE_MS] longer when the row says the subscription renews.
 */
export function isEntitled(row: EntitlementCacheRow | null | undefined, nowMs: number): boolean {
  const until = row?.active_until ? Date.parse(row.active_until) : NaN;
  if (!Number.isFinite(until)) return false;
  const end = row?.will_renew === true ? until + RENEWAL_GRACE_MS : until;
  return end > nowMs;
}

export async function requirePro(admin: Db, userId: string, nowMs: number = Date.now()): Promise<ProCheck> {
  try {
    const { data, error } = await admin
      .from('user_entitlements')
      .select('active_until, period_type, will_renew')
      .eq('user_id', userId)
      .maybeSingle();
    if (error) console.warn('[vana] user_entitlements read failed (treating as not entitled):', error.message);
    else if (isEntitled(data as EntitlementCacheRow | null, nowMs)) return { ok: true };
  } catch (e) {
    console.warn('[vana] user_entitlements read threw (treating as not entitled):', (e as Error).message);
  }
  if (await isAdmin(admin, userId)) return { ok: true };
  return { ok: false, reason: 'pro_required' };
}

/**
 * `public.users.is_admin`, read with the service role (the column is not in
 * the app's cached profile). Any failure, a missing row, `null` or a
 * non-boolean value answers false: the gate fails closed, as the cache read
 * does.
 */
async function isAdmin(admin: Db, userId: string): Promise<boolean> {
  try {
    const { data, error } = await admin.from('users').select('is_admin').eq('id', userId).maybeSingle();
    if (error) {
      console.warn('[vana] users.is_admin read failed (treating as not admin):', error.message);
      return false;
    }
    return (data as { is_admin?: unknown } | null)?.is_admin === true;
  } catch (e) {
    console.warn('[vana] users.is_admin read threw (treating as not admin):', (e as Error).message);
    return false;
  }
}

/**
 * The check every AI function makes at the top of its handler (mp-429 clause 11): null when the caller
 * is entitled, otherwise the refusal to return as is, 403 `{error:'pro_required'}`, the same answer
 * vana-chat gives. Bought credits do not open it; only the RevenueCat cache does.
 */
export async function refuseUnlessPro(admin: Db, userId: string, nowMs: number = Date.now()): Promise<Response | null> {
  const pro = await requirePro(admin, userId, nowMs);
  return pro.ok ? null : jsonResponse({ error: pro.reason }, 403);
}
